lib/internal/modules/esm/loader.js

```diff
@@ -185,6 +203,39 @@ class Loader {
     if (transformSource !== undefined) {
       this._transformSource = FunctionPrototypeBind(transformSource, null);
     }
+    if (getGlobalPreloadCode !== undefined) {
+      this._getGlobalPreloadCode =
+        FunctionPrototypeBind(getGlobalPreloadCode, null);
+    }
+  }
+
+  runGlobalPreloadCode(context = null) {
+    if (!this._getGlobalPreloadCode) {
+      return;
+    }
+    const preloadCode = this._getGlobalPreloadCode();
+    if (preloadCode === null) {
+      return;
+    }
+
+    const { Script } = require('vm');
+    const preloadScript = new Script(preloadCode, {
+      filename: '<global preload code>',
+    });
+    const preloadInit = context ?
+      preloadScript.runInContext(context) :
+      preloadScript.runInThisContext();
```

Comment:
What is wrong with using `eval`? this would also make it easier to inject `getBuiltin` into the eval scope with a contextual eval that carefully avoids exposing anything else.

Reply:
can you clarify what you mean with "contextual eval"? Are you thinking of something like https://nodejs.org/api/vm.html#vm_vm_compilefunction_code_params_options 's contextExtensions params and/or `Function('getBuiltin', code)(getBuiltin)`?

Reply:
Right, just a function with an argument exactly.

Reply:
Now using a function wrapper.